### DISCUSSION TITLE
Is vineyard the only solution for data transfer?

### DISCUSSION THREAD

**Discussion by cfangplus** [alibaba/GraphScope#143] [Q&A] [answered]
Hi,

We know that vineyard is supposed to act as a role for data transfer among the three graph systems and those transferred data is distributed in memory. Also we know that there exists many, many products that could support similar function, such as Redis, Alluxio, or even graph database to some extent could be ok for data transfer among the graph systems.  
So the question is, is vineyard the only solution for this? Could you help to explain what's the real difference between vineyard and the other similar product?

**Comment by sighingnow** [comment, marked as answer]
Hi @cfangplus,

Thanks for your interest on GraphScope!

(I have moved the issue to a Q&A discussion to make it visible for more people.)

> We know that vineyard is supposed to act as a role for data transfer among the three graph systems and those transferred data is distributed in memory.

GraphScope leverages [vineyard][1] to storage large-scale distributed graphs in memory across a cluster. But we DON'T _transfer) data among the three graph systems. Rather, GAE, GIE and GLE in GraphScope shares the graph data in a zero-copy fashion by memory sharing. Those engines refer to the same graph object when performing specific graph computations.

>  many products that could support similar function, such as Redis, Alluxio, or even graph database to some extent could be ok for data transfer among the graph systems.

+ Redis: redis is a KV storage as far as I know. Rather, vineyard address on in-memory sharing for both low level data structures as well 
 as high level data structure abstractions.
+ Alluixo: vineyard isn't a cache layer for filesystem-like storage.
+ GraphScope is not a graph database. GraphScope aims to provides a one-stop graph computation platform for different kinds of workload on large-scale graph data.

> is vineyard the only solution for this?

As described above, we actually don't do data transfer between those engines.

Even when considering scale in/out and the following data migration, vineyard is not the solution for data transfer, rather, the data transfer will be handled by a set of vineyard "drivers" (for more information of the vineyard drivers, please refer to vineyard documentation, and [this discussion thread][2]).

Hope the above could answer you question. Free feel to comment here if you have any other questions 😆 


[1]: https://github.com/alibaba/libvineyard
[2]: https://github.com/alibaba/libvineyard/discussions/157

**Reply by Zarca** [reply]
hi @sighingnow ,
I'm aslo interested  in how the GAE , GIE and GLE sharing only one graph data stored on a vineyard backend.
As far as I know about the project libvineyard, it provide a builder and resolver registration  mechanism.
But I can't find certain builder or resolver that GraphScope provided for this, so I'm  wondering
how GAE and GIE or GLE interpret the same buffers stored on vineyard for different usage ?
Are these engines using different builders and resolvers for their own different purposes?

**Reply by sighingnow** [reply]
You could find [a graph module][1] which includes the distributed graph data structures (i.e., `ArrowFragment`) in libvineyard, which can be directly used by GAE, and has been adapt the [GIE][2] and [GLE][3].

The graph data structure are placed in memory, and mapped to the address space of those engines.

[1]: https://github.com/alibaba/libvineyard/tree/main/modules/graph
[2]: https://github.com/alibaba/GraphScope/tree/main/interactive_engine/src/executor/runtime/native
[3]: https://github.com/alibaba/graph-learn/tree/graphscope/graphlearn/core/graph/storage

**Reply by Zarca** [reply]
Hi,
Thanks for that info, but my concern is how the ArrowFragment been adapted for the GLE or GIE,  perhaps by just modifing the metadata of the payloads in libvineyard?

B.T.W, I 'm also checking through what is the actual gRPC Server's implementation that handles the Operations (like load a graph, add a new column to existing Graph), Cause It seems that the GRPCClient class in:
https://github.com/alibaba/GraphScope/blob/055480bbd4527ec7b7fa434278392f098efaac52/python/graphscope/client/rpc.py#L92
 just supply a Stub (CoordinatorServiceStub) for gRPC of the real Server in k8s .So i wanna to dive into the real server's implementation but i don't get where the real implementation lies.
Look up to your help :)

**Reply by sighingnow** [reply]
> Thanks for that info, but my concern is how the ArrowFragment been adapted for the GLE or GIE, perhaps by just modifing the metadata of the payloads in libvineyard?

They both in C++, thus they reuse the data structure directly. It's also ok to define such fragment type in Python, based on the metadata. That shouldn't be hard.

The code for coordinator service is https://github.com/alibaba/GraphScope/blob/main/coordinator/gscoordinator/coordinator.py

**Reply by sighingnow** [reply]
Currently the "resolver" and "builder" are concepts for vineyard's Python SDK.

**Reply by Zarca** [reply]
Copt that with great appreciation! But still , when I followed the coordinator's codes , it seems that  what the coordinator holds is actual a gRPC stub for the grape engine, too(awesome jobs).
https://github.com/alibaba/GraphScope/blob/ef3df2c82444b677379e3225515a904c949e982c/coordinator/gscoordinator/coordinator.py#L611
 Which means the EngineServiceStub Class 's server end do the real jobs or Operation (such as Create a graph, doing graph computations and , etc. ) for us. Is that right?
If it is, i still wanna to know where the EngineServiceStub's corresponding implmentation lies, plz.

**Reply by Zarca** [reply]
For the second quesiotn ,If the builder or resolver is not used (as you said they are Python SDKs) , then how do the GAE and GLE recognize the object stored in vineyard? Just set different metas while keeping the same buffers(Blobs ) under the hood?

**Reply by sighingnow** [reply]
> then how do the GAE and GLE recognize the object stored in vineyard

They depends on the vineyard's C++ library.

> If it is, i still wanna to know where the EngineServiceStub's corresponding implmentation lies

Here you go: https://github.com/alibaba/GraphScope/blob/main/analytical_engine/core/grape_instance.cc

**Reply by Zarca** [reply]
Got it, but i'm not very familiar with C++ , and i also found that the GIE engine is  written in Java,
So it puzzled me again how could the GIE engine utilize the libvineyard? (as far as i know, the vineyard hasn't supplied a Java SDK, currently)

**Reply by sighingnow** [reply]
The core GIE execution engine is written in Rust, and there is a FFI bindings to read fragment in vineyard: https://github.com/alibaba/GraphScope/tree/main/interactive_engine/src/executor/runtime/native

**Reply by Zarca** [reply]
Okay，now it make sense :)
However, after  checking through the codes you said:
> You could find [a graph module](https://github.com/alibaba/libvineyard/tree/main/modules/graph) which includes the distributed graph data structures (i.e., `ArrowFragment`) in libvineyard, which can be directly used by GAE, and has been adapt the [GIE](https://github.com/alibaba/GraphScope/tree/main/interactive_engine/src/executor/runtime/native) and [GLE](https://github.com/alibaba/graph-learn/tree/graphscope/graphlearn/core/graph/storage).
>
I supposed that  it means the Graph of GAE is stored as vineyard:NumericArray and vineyard:Table ?  If it is ,
could you please show me where and how the GLE resolve this Graph (created by GAE) through libvineyard's C++  API?

**Reply by Zarca** [reply]
cause i thinks  it is quite difficult for me to walk though such a complex project , i just wanna get  how the Graph defined by GAE  is interpreted or resolved by GLE engine, explicitly

**Reply by sighingnow** [reply]
Please refer to code in above links for the detail implementation.

**Reply by Zarca** [reply]
ok, now i have seen how the header: arrow_fragment.h is included by those three engines, 
So ,here comes my most puzzled concern: now that GLE has used arrow_fragment.cc to store or interpreter the graph, Did it use the python' s resolver to resolve the graph before getting the torch or tensorflow computation?

**Reply by sighingnow** [reply]
The learning engine incluedes two part: server and client.

+ the server part is in C++ and responsible for reading graph and do sampling
+ the client part is in python and request for data from server using gRPC.

The client python (the python part) has nothing to do with the underlying distributed graph.

**Reply by sighingnow** [reply]
For more details about the learning engine, please refer to https://github.com/alibaba/graph-learn.

**Reply by Zarca** [reply]
> The learning engine incluedes two part: server and client.
> 
> * the server part is in C++ and responsible for reading graph and do sampling
> * the client part is in python and request for data from server using gRPC.
> 
> The client python (the python part) has nothing to do with the underlying distributed graph.

So, the client calls gRPC to vineyard with an ObjectID to get the Data back , and then use resolver to resolve the vineyard::NumericArray and etc... (C++ data ) to Python object (Python data) , then start a  torch or tensorflow backend to do the GNN stuffs,  Is that right?
* And BTW, It seems that the distributed graph 's  all partitions in GAE or GIE on different machines can be computed totally alone, without any data shuffle or exchange, but in my opinion, many Algorithms (like PageRank) will have to get messages 
from all neighbor nodes in the Graph , how could GAE handle it when data are stored separately on different vineyard instances?(since The vineyard hasn't supported data migration as i knew)

hope not a bother for you :)

**Reply by sighingnow** [reply]
> the client calls gRPC to vineyard with an ObjectID to get the Data back

Not to vineyard, it is to learning engine server.

> and then use resolver to resolve the vineyard::NumericArray and etc... (C++ data ) to Python object (Python data) 

The learning engine has nothing to do with the python SDK of vineyard, its server use vineyard's C++ SDK, and the client part doesn't depends on vineyard package.

>  how could GAE handle it when data are stored separately on different vineyard instances?

GAE itself is built on top of MPI and has its own message system. For the core-engine of GAE, please refer to https://github.com/alibaba/libgrape-lite

**Reply by Zarca** [reply]
Thanks for it @sighingnow , dit the GIE and GLE all have their own MPI system for data exchange?

**Reply by sighingnow** [reply]
+ GIE has its own dataflow model and implements cross worker communication, for data exchange, not on MPI. 
+ GLE uses gRPC for such functionalities.

**Reply by Zarca** [reply]
It helps a lot, now i got understood the things under the hood 
👍

**Reply by sighingnow** [reply]
> It helps a lot, now i got understood the things under the hood
> 👍

Thanks for being interested on GraphScope!

**Reply by Zarca** [reply]
BTW, I'm puzzled about missing not only the grape/config.h but also something like the grape::InitMPIComm()etc. after cloning the Graph using "git clone ..." and "git submodule --init".
Maybe  they are all defined outside this GraphScope's project?

**Reply by sighingnow** [reply]
> > how could GAE handle it when data are stored separately on different vineyard instances?
> 
> GAE itself is built on top of MPI and has its own message system. For the core-engine of GAE, please refer to https://github.com/alibaba/libgrape-lite

GraphScope dependes on libgrape-lite, you need to install libgrape-lite first.

**Reply by Zarca** [reply]
Got it ,and again , as for the question above:

> * GLE uses gRPC for such functionalities.
>
However, the official example which the line  showed in :https://github.com/alibaba/GraphScope/blob/main/demo/node_classification_on_citation.ipynb
lg = sess.learning(sub_graph, nodes=[("paper", paper_features)],
                   edges=[("paper", "cites", "paper")],
                   gen_labels=[
                       ("train", "paper", 100, (0, 75)),
                       ("val", "paper", 100, (75, 85)),
                       ("test", "paper", 100, (85, 100))
                   ])

seems to construct a graph-learn graph ,  but with only one worker?
My concern is  now that the graph is stored in a distributed manner in vineyard on different machines, so why would the graph-learn engine has just one instance ?

**Reply by Zarca** [reply]
I think it is a client process to communicate with graph-learn's server processes in k8s pods,
Is that true?

**Reply by sighingnow** [reply]
> I think it is a client process to communicate with graph-learn's server processes in k8s pods,

That works indeed in such way.

> why would the graph-learn engine has just one instance ?

The graph-learn instances are launched in **every** pod of graphscope. In our demo, there are only one client, which will talk to all of the graph-learn server instances.

**Reply by Zarca** [reply]
Dose every graph-learn's server's instance all need to load a full graph stored on vineyard (using gRPC) ?
Or the GLE just go straight doing a distributed training ?( i.e. each server compute its own partition, like what you told me )
Cause I found the dist demo of GraphSage showed in:
https://github.com/alibaba/graph-learn/blob/master/examples/tf/graphsage/dist_train.py
 seem to implys that each GLE server must load a full graph rather just a partition before the cluster's training,
which seems to be much different  from what graphScope are doing here.

**Reply by sighingnow** [reply]
> Dose every graph-learn's server's instance all need to load a full graph stored on vineyard (using gRPC) ?

Each server compute its own partition.

> which seems to be much different from what graphScope are doing here.

For distributed training, every client will be mapped to a subset of server.

**Reply by Zarca** [reply]
Cool and awesome C++ project!
After checking the core data loading process of GAE in en_fragment_loader:
https://github.com/alibaba/libgrape-lite/blob/7903afded169ead2a7cd88cef71130d0d75f4b97/grape/fragment/ev_fragment_loader.h#L101

 I assume that all GAE server instances will need a complete full vertex file, and a partition of the edge file(even split) . Then it can hold a subgraph of the full graph (all nodes, but part edges ) , and began the algorithm like SSSP,PageRank in a parallelized way , which is discussed in the paper: 
https://dl.acm.org/doi/pdf/10.1145/3035918.3035942.
But, if this is the true, i still get struggled about how can the subgraph inside each of the GAE's instance be recognized by GLE? It looks like that the GLE must have the same ability to treat a distributed subgraphs like GAE.
(which i think can't  be handled by just using the same set of C++ API of vineyard of ArrowFragment)

**Reply by sighingnow** [reply]
> all nodes, but part edges

Not all nodes. Nodes are partitioned using hash.

> GAE's instance be recognized by GLE

GLE does have the the ability to processing distributed graphs.

> i think can't be handled by just using the same set of C++ API of vineyard of ArrowFragment

Why cannot ?

**Reply by Zarca** [reply]
Hi,
 i do check again the core loading process, and now i believe it is indeed every node has been hashed and assigned to different fragments.
As for this puzzle:
> > i think can't be handled by just using the same set of C++ API of vineyard of ArrowFragment
> 
> Why cannot ?

What I'm trying to say is that, considered we have a GAE's distributed graph here(using MPI to send nodes and edges to the corresponding fragment's machine). But when it comes to  GLE, unless the GLE itself has exact the same message system design  as the GAE (and exact the same graph building way like GAE's ev_fragment_loader do), otherwise, I think it can't do computation on the Graph DataStructure created by GAE directly.

For instance, if the GLE is aim to launch a GCN , which will aggregate the node's neighbors from other machines,
must it call some API from GAE to get such message from remote subgraph(or the so called "fragment")?

Off course for me, I think this part is one of the key parts of the whole GraphScope's brilliant innovation, but i also think it is difficult to understand. 

Cause I'm thinking how i can  adapt some third party's GAE or GLE or GIE to shared the same distributed graph even the message system, like you did in GraphScope.

**Reply by Zarca** [reply]
In a word, i.e. how to adapt a distributed graph defined and split by GAE's logics to another engine (like GLE )？
For instance, the distributed graph may have nodes hashed to different machine, so if a  GLE engine wanna to reuse this graph, it's partition logics must be the very same as GAE .
Is it that true? (GLE and GAE have exactly the same graph partition strategy ?, or the GLE has to repartition a new graph from the GAE's output?)

**Reply by sighingnow** [reply]
> GLE and GAE have exactly the same graph partition strategy ?

They shares the same graph data, so, the partition is same as well.

**Comment by Zarca** [comment]
Do GLE also follows the same distributed graph model's communication mode (i.e the same partition strategy, the PEval, the IncEva and Assemble communication mode ) as GAE(GRAPE)'s ?
Or, the GAE can simulate another parallel model of other engines, without a reconstruction of graph?（like Grape's paper said）

**Reply by sighingnow** [reply]
> Do GLE also follows the same distributed graph model's communication mode (i.e the same partition strategy, the PEval, the IncEva and Assemble communication mode ) as GAE(GRAPE)'s ?

No, GLE has own definition for distributed communication.

>（like Grape's paper said）

GLE has no link with grape's paper.

**Reply by Zarca** [reply]
Got it, they just share the same graph partition strategy(like edge cut or something else...), but the communication systems are totally independent from each others. So did the gremline engine (GIE) also can share and recognize the same graph partition (like GAE and GLE did)?
And if any third party graph engine wanna to work with GraphScope, I think it must be a necessity that  the third party engine should share a same distributed graph partition with the three engines of GraphScope, is that true? 
Otherwise, it has to rebuild the graph itself.

**Reply by sighingnow** [reply]
> So did the gremline engine (GIE) also can share and recognize the same graph partition (like GAE and GLE did)?

Definitely yes.

> Otherwise, it has to rebuild the graph itself.

The answer is yes as well. But we promise our implementation for graph data structure is highly-optimized and yields pretty good performance.

**Reply by Zarca** [reply]
Thanks for all the kind and patient relies , now i got know  the awesome things and the data flow outlines  behind GraphScope :)

**Reply by Zarca** [reply]
B.T.W. , Finally, I suppose the GAE's graph partition strategy is a edge cut way by default,（through MPI）,
Is that true? can a vertex cut strategy be applied to GraphScope?

**Reply by sighingnow** [reply]
Currently the fragment is partitioned as edge cut. Vertex cut will be released later (may before Sep, 2021).

**Reply by Zarca** [reply]
Got it 🥇
